core/src/main/scala/org/ensime/core/AnalyzerManager.scala

```diff
@@ -0,0 +1,171 @@
+    case req @ RestartScalaCompilerReq(id, _) =>
+      id match {
+        case Some(projectId) =>
+          analyzers.get(projectId).foreach(_ forward req)
+        case None if analyzers.nonEmpty =>
+          analyzers.values foreach (_ forward req)
+        case None =>
+          broadcaster ! AnalyzerReadyEvent
+      }
+    case req @ UnloadAllReq =>
+      analyzers.foreach {
+        case (_, analyzer) => analyzer forward req
+      }
+    case req @ TypecheckModule(moduleId) =>
+      getOrSpawnNew(moduleId) forward req
+    case req @ RemoveFileReq(file: File) =>
+      val f = SourceFileInfo(RawFile(file.toPath), None, None)
+      withExistingModuleFor(f, req)((req, moduleId) =>
+        getOrSpawnNew(moduleId) forward req)
+    case req @ TypecheckFileReq(fileInfo) =>
+      withExistingModuleFor(fileInfo, req)((req, moduleId) =>
+        getOrSpawnNew(moduleId) forward req)
+    case req @ TypecheckFilesReq(files) =>
+      if (files.exists(config.findProject(_).isEmpty))
+        sender ! EnsimeServerError("Update .ensime file.")
+      else {
+        val original = sender
+        val filesPerProject = files.groupBy(config.findProject(_)).map(x => x._1.get -> x._2)
+
+        context.actorOf(Props(new Actor {
+          private var remaining = filesPerProject.size
+          private var aggregate: List[String] = List.empty
+
+          override def preStart: Unit =
+            for ((moduleId, list) <- filesPerProject)
+              getOrSpawnNew(moduleId) ! TypecheckFilesReq(list)
+
+          override def receive = {
+            case res: RpcResponse if remaining > 1 =>
+              aggregate = addResponse(res, aggregate)
+              remaining -= 1
+            case res: RpcResponse =>
+              aggregate = addResponse(res, aggregate)
+              original ! combine(aggregate)
+              context.stop(self)
+          }
+
+          def addResponse(res: RpcResponse, agg: List[String]) = res match {
+            case EnsimeServerError(desc) =>
+              desc :: aggregate
+            case _ =>
+              aggregate
+          }
+
+          def combine(errors: List[String]): RpcResponse =
+            if (aggregate.isEmpty) // had no errors; return a  VoidResponse
+              VoidResponse
+            else // return the cumulative error
+              EnsimeServerError(aggregate mkString ", ")
+        }))
+      }
+    case req @ RefactorReq(_, _, _) =>
+      val original = sender
+      context.actorOf(Props(new Actor {
+        override def preStart(): Unit = {
+          context.actorOf(analyzerCreator(config.projects.map(_.id))) ! req
+        }
+        override def receive = {
+          case res: RpcResponse =>
+            original ! res
+            context.stop(self)
+        }
+      }))
+    case req @ CompletionsReq(fileInfo, _, _, _, _) =>
+      withExistingModuleFor(fileInfo, req)((req, moduleId) =>
+        getOrSpawnNew(moduleId) forward req)
```

Comment:
this pattern of getting the analyzer and forwarding the request seems like it could be abstracted instead of repeated everywhere.